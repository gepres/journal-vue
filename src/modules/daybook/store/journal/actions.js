import journalApi from "@/api/journalApi"

export const loadEntries = async ({commit}) => {
  const {data} = await journalApi.get('/entries.json')
  if(!data){
    commit('setEntries',[])
    return
  }
  const entries = []
  for (const id of Object.keys(data)) {
    entries.push({
      id,
      ...data[id]
    })
  }
  commit("setEntries",entries)
}

export const updateEntry = async ({commit},item) => {
  // quitar el id   delete item['id'];
  const { date, picture, text } = item
  const dataToSave = { date, picture, text }
  await journalApi.put(`/entries/${item.id}.json`,dataToSave)
  dataToSave.id = item.id
  commit('updateEntry',{...dataToSave})
}

export const createEntry = async ({commit}, item) => {
  const { date, picture, text } = item
  const dataToSave = { date, picture, text }
  const {data} = await journalApi.post(`/entries.json`,dataToSave)
  dataToSave.id = data.name
  commit('addEntry',dataToSave)
  return {id:data.name}
}

export const deleteEntry = async ({commit},id) => {
  await journalApi.delete(`/entries/${id}.json`)
  commit('deleteEntry',id)
}