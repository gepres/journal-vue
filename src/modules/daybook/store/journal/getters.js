export const getEntriesByTerm = (/*state*/) => {
}

export const GetEntryById = (/*state*/) => {
}