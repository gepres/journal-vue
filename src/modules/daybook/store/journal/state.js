export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Delectus consectetur asperiores qui! Assumenda, eius, numquam adipisci fugiat qui neque nesciunt rem natus quisquam tempora officiis expedita dicta optio, nobis ab!',
      picture: null
    },
    {
      id: new Date().getTime()+1000,
      date: new Date().toDateString(),
      text: 'Lorem ipsum dolors expedita dicta optio, nobis ab!',
      picture: null
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quibusdam, aliquid? Eum nobis, enim natus quia praesentium ullam ipsum recusandae quibusdam est saepe harum cupiditate deserunt asperiores ratione nostrum temporibus similique eius ex quis libero maxime atque. Omnis delectus nisi itaque!',
      picture: null
    }
  ]
})
