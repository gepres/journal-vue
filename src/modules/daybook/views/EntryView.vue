<template>
  <template v-if="entry && !loading">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{day}}</span>
        <span class="mx-1 fs-3">{{month}}</span>
        <span class="mx-2 fs-4 fw-light">{{yearDay}}</span>
      </div>
      <div>
        <input v-show="false" ref="imageSelector" type="file" @change="onSelectedImage($event)" accept="image/*">
        <button v-if="entry.id" @click="onDeleteEntry" class="btn btn-danger mx-2">
          Borrar
          <i class="fa fa-trash-alt mx-1" />
        </button>
        <button @click="onSelectImage"  class="btn btn-primary">
          Subir foto
          <i class="fa fa-upload mx-1" />
        </button>
      </div>
    </div>
    <hr>
    <div class="d-flex flex-column px-3 h-75">
      <textarea v-model="entry.text" placeholder="¿ Que sucedio hoy?"></textarea>
    </div>
    <!-- <img class="img-thumbnail" src="https://www.solofondos.com/wp-content/uploads/2015/04/Fondos-de-paisajes.jpg" alt="paisaje"> -->
     <img v-if="entry.picture && !localImage" class="img-thumbnail" :src="entry.picture" alt="">
    <img v-if="localImage" class="img-thumbnail" :src="localImage" alt="">
  </template>
  <Loading v-if="loading" />
  <Fab v-if="!loading" icon="fa-save" @fabClick="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from '@vue/runtime-core'
import { mapGetters, mapActions } from 'vuex'
import Swal from 'sweetalert2'
import getDayMothYear from '../helpers/getDayMothYear'
import uploadImage from '../helpers/uploadImage'
export default {
  name:'entryView',
  props:{
    id: {
      type: String,
      required: true
    }
  },
  components: {
    Fab: defineAsyncComponent(() => import('@/modules/daybook/components/FabComponent.vue')),
    Loading: defineAsyncComponent(() => import('@/modules/shared/components/loading.vue'))
  },
  data () {
    return {
      entry:null,
      loading: false,
      localImage: null,
      file: null
    }
  },  
  created(){
    this.loadEntry()
  },
  watch:{
    id(){
      this.loadEntry()
    }
  },
  methods: {
    ...mapActions({
      updateEntry: 'journal/updateEntry',
      createEntry: 'journal/createEntry',
      deleteEntry: 'journal/deleteEntry'
    }),
    loadEntry(){
      let entry
      if(this.id === 'new'){
        entry = {
          text: '',
          date: new Date().getTime()
        }
      }else{
        entry = this.GetEntryById(this.id)
        if(!entry) return this.$router.push({name:'no-entry'})
        this.localImage = entry.picture
      }
      
      this.entry = entry
    },
   async saveEntry (){
     this.loading = true
     Swal.fire({
       title: 'Espere por favor',
       allowOutsideClick: false
     })
     Swal.showLoading()
     const picture = await uploadImage(this.file)
     this.entry.picture = picture
     if(this.entry.id){
        await this.updateEntry({...this.entry})
     } else {
      //  crear una nueva entra
      const {id} = await this.createEntry({...this.entry})
        this.$router.push({name: 'entry', params: {id}})
     }

     this.loading = false
     this.file = null
     Swal.fire('Guardado', 'Entrada registrada con éxito','success')
    },
    async onDeleteEntry() {
      console.log('se llamo aqui');
      this.loading = true
      const {isConfirmed} = await Swal.fire({
        title: '¿ Está seguro ?',
        text: 'una ves borrado no se puede recuperar',
        showDenyButton: true,
        confirmButtonText: 'Si, estoy seguro'
      })
      console.log(isConfirmed);
      if(isConfirmed) {
        Swal.fire({
          title: 'Espere por favor',
          allowOutsideClick: false
        })
        Swal.showLoading()
        console.log('apúnto de eliminar');
        await this.deleteEntry(this.entry.id)
        this.loading = false
        this.$router.push({name: 'no-entry'})
        Swal.fire('Eliminado', '', 'success')
      } else {
        this.loading = false
      }
    },
    onSelectedImage($event) {
      const file = $event.target.files[0]
      if (!file) {
        this.localImage = null
         this.file = null
        return
      } 
      this.file = file
      const fr = new FileReader()
      fr.onload = () => this.localImage = fr.result
      fr.readAsDataURL(file)
    },
    onSelectImage() {
      this.$refs.imageSelector.click()
    }
  },
  computed: {
    ...mapGetters({
      GetEntryById: 'journal/GetEntryById',
    }),
    day(){
      const {day} = getDayMothYear(this.entry.date)
      return day
    },
    month(){
      const {month} = getDayMothYear(this.entry.date)
      return month
    },
    yearDay() {
      const {yearDay} = getDayMothYear(this.entry.date)
      return yearDay
    }
  },
}
</script>

<style lang="scss" scoped>
textarea{
  font-size: 20px;
  border:none;
  height: 100%;
  &:focus{
    outline: none;
  }
}
img{
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow:0px 5px 10px rgba($color:#000, $alpha: 0.2);
}
</style>