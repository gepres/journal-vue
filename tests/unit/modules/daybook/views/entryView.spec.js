import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import Swal from 'sweetalert2'
import EntryView from '@/modules/daybook/views/EntryView'
import journal from '@/modules/daybook/store/journal';
import { journalState } from "../../../mock-data/test-journal-state";


const createVuexStore = (initialState) => createStore({
  modules: {
    journal: {
      ...journal,
      state: {...initialState}
    }
  }
})

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn()
}))

describe('Pruebas en el Entry view', () => {

  const store = createVuexStore( journalState)
  store.dispatch = jest.fn()
  // store.mutations = jest.fn()
  const mockRouter = {
    push: jest.fn()
  }
  let wrapper
  beforeEach(() => {
    jest.clearAllMocks()
    wrapper = shallowMount(EntryView,{
      props: {
        id: '456qwertyu'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins:[store]
      }
    })
  })
  


  test('debe de sacar al usuario porque el id no existe', () => {
    const wrapper = shallowMount(EntryView,{
      props: {
        id: 'Este ID no existe'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins:[store]
      }
    })
    expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
  });


  test('debe de mostrar la entrada correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(mockRouter.push).not.toHaveBeenCalled()
  });


  test('debe de borrar la entrada y salida', (done) => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true }))
    wrapper.find('.btn-danger').trigger('click')
    expect(Swal.fire).toHaveBeenCalledWith({
      title: '¿ Está seguro ?',
      text: 'una ves borrado no se puede recuperar',
      showDenyButton: true,
      confirmButtonText: 'Si, estoy seguro'
    })
    setTimeout(() => {
      // expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '123asdfghjkl')
      expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "456qwertyu")
      expect(mockRouter.push).toHaveBeenCalled()
      done()
    }, 1);
  });
});