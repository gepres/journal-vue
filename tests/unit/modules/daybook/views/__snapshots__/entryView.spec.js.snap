// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Entry view debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">26</span><span class="mx-1 fs-3">Mayo</span><span class="mx-2 fs-4 fw-light">2022, Jueves</span></div>
  <div><input type="file" accept="image/*" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt mx-1"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload mx-1"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Â¿ Que sucedio hoy?"></textarea></div>
<!-- <img class="img-thumbnail" src="https://www.solofondos.com/wp-content/uploads/2015/04/Fondos-de-paisajes.jpg" alt="paisaje"> -->
<!--v-if-->
<!--v-if-->
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
